[SPARK-42038][SQL] SPJ: Support partially clustered distribution

sunchao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

Currently with [storage-partitioned join](https://docs.google.com/document/d/1foTkDSM91VxKgkEcBMsuAvEjNybjja-uHk-r3vtXWFE/edit#heading=h.82w8qxfl2uwl), both sides of the join must be **fully clustered** over the partition values, that is, each Spark partition should have a distinct partition value. To guarantee this, Spark will group all the input partitions reported by a V2 data source on the partition values. The consequence, however, is that it can easily lead to data skew, when a few partitions have large amount of data.

This PR introduce a new mechanism which requires only one side of the storage-partitioned join to be fully clustered, while the other side can be **partially clustered**, i.e., there could exist multiple Spark partitions with the identical partition value. At planning time, Spark will compare the statistics from both sides of the join, and pick the side with smaller size as full clustered side, while the other side is partially clustered. It then replicate the partitions on the former to match the number of partitions on the latter. 

The concept of this optimization is similar to the existing `OptimizeSkewedJoin`, and techniques such as key salting, but is applied on DataSource V2 level instead, and doesn't require AQE to be enabled. Unlike `OptimizeSkewedJoin`, however, this optimization is applied before any shuffle happens.

A new config `spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled` is added to enable & disable the feature. By default it is turned off.

For instance, consider the following SQL query:
```sql
SELECT * FROM a JOIN b on a.id = b.id
```

while table `a` reports partitions `[0, 1, 2]` while `b` reports partitions `[0, 1, 1, 1, 2, 2]`.

Without the PR, Spark would group input partitions so that both sides will have 3 partitions `[0, 1, 2]`. With the PR, Spark will choose the right-hand side as partially clustered and match the left-hand side with it. Therefore, both sides have 6 partitions `[0, 1, 1, 1, 2, 2]`. 

Note this PR currently relies on a simple heuristic and always pick the side with less data size based on table statistics as the side fully clustered, even though it could also contain skewed partitions. In future, we can potentially do fine-grained comparison based on partition values.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

As mentioned in the previous section, this feature can help to reduce data skewness during storage-partitioned join, when a few partitions are mapped to a large amount of data.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

A new Spark config, `spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled`, is introduced to enable/disable the feature. By default it is disabled, so the behavior should still be the same as before.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

Added new tests in `KeyGroupedPartitioningSuite`.

sunchao committed:
initial commit

HyukjinKwon commented:
@sunchao do you plan to add this for 3.4? cc @xinrong-meng FYI

sunchao commented:
@HyukjinKwon it'd be great if this can make it into 3.4, but it's not a release blocker.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1424,6 +1424,18 @@ object SQLConf {
+            "statistics, group and replicate them to match the other side. This can help to " +
+            "reduce data skewness when certain partition values have large amount of data. This " +
+            s"config requires both ${V2_BUCKETING_ENABLED.key} and " +
+            s"${V2_BUCKETING_PUSH_PART_VALUES_ENABLED.key} to be enabled")
```
We need a version.
```scala
.version("3.4.0")
```

dongjoon-hyun reviewed (approved):
+1, LGTM (with only one minor comment about config version). Thank you so much, @sunchao .

cc @aokolnychyi , @viirya , @huaxingao , @cloud-fan

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1424,6 +1424,18 @@ object SQLConf {
 
+  val V2_BUCKETING_PARTIALLY_CLUSTERED_DISTRIBUTION_ENABLED =
+    buildConf("spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled")
+        .doc("During a storage-partitioned join, whether to allow input partitions to be " +
```
Indentation?

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:
```diff
@@ -242,8 +246,16 @@ abstract class InMemoryBaseTable(
       dataMap += dataMap.get(key)
-        .map(key -> _.withRow(row))
-        .getOrElse(key -> new BufferedRows(key).withRow(row))
+          .map { splits =>
```
Indentation?

cloud-fan commented:
Are we replicating the AQE skew join optimization in `EnsureRequirements`? I'm a bit worried about direction as `EnsureRequirements` is becoming more and more complicated.

cloud-fan commented:
Since the storage partitioned join becomes more and more complicated, shall we revisit the framework? Essentially we need to join the partitions from two v2 tables at the driver side, and launch a task to process each partition pair. One idea is to add a physical rule before `EnsureRequirements`. This new rule matches join of two v2 tables, applies a join strategy, and sets a special flag in the join node so that `EnsureRequirements` can skip it. This is similar to `ShuffledJoin.isSkewJoin`:
```
trait ShuffledJoin extends JoinCodegenSupport {
  def isSkewJoin: Boolean

    override def requiredChildDistribution: Seq[Distribution] = {
    if (isSkewJoin) {
      // We re-arrange the shuffle partitions to deal with skew join, and the new children
      // partitioning doesn't satisfy `HashClusteredDistribution`.
      UnspecifiedDistribution :: UnspecifiedDistribution :: Nil
    } else {
      ClusteredDistribution(leftKeys) :: ClusteredDistribution(rightKeys) :: Nil
    }
  }
}
```

sunchao commented:
@cloud-fan the idea is similar to skew join but for v2 sources, let me try to split the code into a separate rule following your idea.

sunchao commented:
Hmm @cloud-fan , it seems a bit tricky to split the logic specific to SPJ out of `EnsureRequirements`, since it depends on `reorderJoinPredicates`, which is currently co-mingled with `EnsureRequirements`. I was trying to split `reorderJoinPredicates` into a separate rule but just realized I hit [SPARK-22042](https://issues.apache.org/jira/browse/SPARK-22042).

sunchao committed:
revert

dongjoon-hyun commented:
Shall we convert to `Ready for review` status back, @sunchao ?

sunchao committed:
more test

sunchao commented:
Yes, converted.

I found it is quite difficult to move the logic out of `EnsureRequirements`, because, as mentioned above, the optimization also depends on `reorderJoinPredicates` which needs to be executed together with `ensureDistributionAndOrdering` in a lock-step as we go up the query plan.

In addition, even if I extract out `reorderJoinPredicates` into a separate util method or something, it's still difficult to make the optimization as a separate rule because it also depends on certain part of the logic in `ensureDistributionAndOrdering` (for instance, as we go up the query plan tree, we'd expect join branches that have incompatible `KeyGroupedPartitioning`s to be handled by `ensureDistributionAndOrdering` and converted to use hash partitioning, so that we should not try to apply the optimization again on those branches).

As a compromise, I've extracted all special logic related to `KeyGroupedPartitioning` into a separate method `checkKeyGroupCompatible`, so the main body of `ensureDistributionAndOrdering` now looks much simpler. If necessary, I can also move this method (and related) into a separate class file instead, to make them more isolated.

dongjoon-hyun commented:
Thank you, @sunchao !

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1432,6 +1432,19 @@ object SQLConf {
+        "on skew join and can help to reduce data skewness when certain partitions are assigned " +
+        s"large amount of data. This config requires both ${V2_BUCKETING_ENABLED.key} and " +
+        s"${V2_BUCKETING_PUSH_PART_VALUES_ENABLED.key} to be enabled")
+      .version("3.4.0")
```
If you don't mind, we still want to have this in 3.4.0, @cloud-fan and @xinrong-meng . Please let us know your opinions.

sunchao committed:
improve test

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -40,15 +41,18 @@ case class BatchScanExec(
       this.batch == other.batch && this.runtimeFilters == other.runtimeFilters &&
-          this.commonPartitionValues == other.commonPartitionValues
+          this.commonPartitionValues == other.commonPartitionValues &&
+          this.replicatePartitions == other.replicatePartitions
```
shall we also check `applyPartialClustering`?

sunchao replied:
Oops good catch

sunchao committed:
rebase and address comments

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:
```diff
@@ -411,6 +336,235 @@ case class EnsureRequirements(
+    if (isCompatible) Some(Seq(newLeft, newRight)) else None
+  }
+
+  // Similar to `OptimizeSkewedJoin.canSplitRightSide`
```
This comment looks not informative.

sunchao replied:
I wanted to point out these two methods are identical to the ones in `OptimizeSkewedJoin` except the method name

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:
```diff
@@ -411,6 +336,235 @@ case class EnsureRequirements(
+          logInfo("Calculating partially clustered distribution for " +
+              "storage-partitioned join")
+
+          val canReplicateLeft = canReplicateLeftSide(joinType)
+          val canReplicateRight = canReplicateRightSide(joinType)
```
Might be good to add a comment for disallowed cases (specified sides in outer join) and the reason.

sunchao replied:
+1. Added comments for this.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -132,12 +140,66 @@ case class BatchScanExec(
-          finalPartitions = p.partitionValues.map { partValue =>
-            // Use empty partition for those partition values that are not present
-            partitionMapping.getOrElse(partValue, Seq.empty)
+          // This means the input partitions are not grouped by partition values. We'll need to
+          // check `groupByPartitionValues` and decide whether to group and replicate splits within
+          // a partition.
```
Does this comment need to be moved below on top of `if (commonPartitionKeys.isDefined ...`?

sunchao replied:
you're right - moved.

sunchao committed:
more comments

dongjoon-hyun commented:
Could you resolve the conflicts once more, @sunchao ?

sunchao committed:
fix InternalRow comparability

dongjoon-hyun commented:
Merged to master/3.4.

Thank you so much, @sunchao , @cloud-fan , @viirya .

Also, cc @aokolnychyi and @RussellSpitzer

sunchao commented:
Thanks @cloud-fan @dongjoon-hyun @viirya !

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -39,15 +40,19 @@ case class BatchScanExec(
-    commonPartitionValues: Option[Seq[InternalRow]] = None) extends DataSourceV2ScanExecBase {
+    commonPartitionValues: Option[Seq[(InternalRow, Int)]] = None,
+    applyPartialClustering: Boolean = false,
+    replicatePartitions: Boolean = false) extends DataSourceV2ScanExecBase {
```
This happens quite a few times that we add a new parameter to `BatchScanExec` and change a lot of code which does a pattern matching over `BatchScanExec`. Can we put SPJ-related information in a wrapper class? Then `BatchScanExec` can be more stable. cc @sunchao

sunchao replied:
@cloud-fan yes, it's a good idea. The issue is quite annoying. Let me create a PR for it.

anigos commented:
Partially Clustered Distribution doesn't work with Full Outer Joins for example when we are doing merge. In this cases it produces redundant dataset. Is there any development going on for that? @sunchao @viirya @dongjoon-hyun
